decide.go

```diff
@@ -19,10 +19,137 @@ type DecideRequestData struct {
 	GroupProperties  map[string]Properties `json:"group_properties"`
 }
 
+// FlagDetail represents a feature flag in v4 format
+type FlagDetail struct {
+	Key      string       `json:"key"`
+	Enabled  bool         `json:"enabled"`
+	Variant  *string      `json:"variant"`
+	Reason   *FlagReason  `json:"reason"`
+	Metadata FlagMetadata `json:"metadata"`
+}
+
+// FlagReason represents why a flag was enabled/disabled
+type FlagReason struct {
+	Code           string `json:"code"`
+	Description    string `json:"description"`
+	ConditionIndex int    `json:"condition_index"`
+}
+
+// FlagMetadata contains additional information about a flag
+type FlagMetadata struct {
+	ID          int     `json:"id"`
+	Version     int     `json:"version"`
+	Payload     *string `json:"payload"`
+	Description *string `json:"description,omitempty"`
+}
+
+// GetValue returns the variant if it exists, otherwise returns the enabled status
+func (f FlagDetail) GetValue() interface{} {
+	if f.Variant != nil {
+		return *f.Variant
+	}
+	return f.Enabled
+}
+
+// NewFlagDetail creates a new FlagDetail from a key, value, and optional payload
+func NewFlagDetail(key string, value interface{}, payload *string) FlagDetail {
+	var variant *string
+	var enabled bool
+
+	switch v := value.(type) {
+	case string:
+		variant = &v
+		enabled = true
+	case bool:
+		enabled = v
+	default:
+		enabled = false
+	}
+
+	return FlagDetail{
+		Key:     key,
+		Enabled: enabled,
+		Variant: variant,
+		Reason:  nil,
+		Metadata: FlagMetadata{
+			Payload: payload,
+		},
+	}
+}
+
+// DecideResponse represents the response from the decide endpoint v3 or v4.
+// It is a normalized super set of the v3 and v4 formats.
 type DecideResponse struct {
+	CommonResponseFields
+
+	// v4 flags format
+	Flags map[string]FlagDetail `json:"flags,omitempty"`
+
+	// v3 legacy fields
 	FeatureFlags        map[string]interface{} `json:"featureFlags"`
 	FeatureFlagPayloads map[string]string      `json:"featureFlagPayloads"`
-	QuotaLimited        *[]string              `json:"quota_limited"`
+}
+
+// CommonResponseFields contains fields common to all decide response versions
+type CommonResponseFields struct {
+	QuotaLimited              *[]string `json:"quota_limited"`
+	RequestId                 string    `json:"requestId"`
+	ErrorsWhileComputingFlags bool      `json:"errorsWhileComputingFlags"`
+}
+
+// UnmarshalJSON implements custom unmarshaling to handle both v3 and v4 formats
+func (r *DecideResponse) UnmarshalJSON(data []byte) error {
```

Comment:
this is a great abstraction, very idiomatic golang

Reply:
Thanks!